File: backend/app/schemas/gerencia.py
"""Schemas Pydantic do módulo Gerência (monday). Só saída (Out) + patch de board."""
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel


class BoardOut(BaseModel):
    id: str
    monday_board_id: int
    nome: str
    workspace: str | None = None
    board_kind: str | None = None
    incluido: bool
    confianca_classificacao: str | None = None
    ativo: bool
    colunas_map: dict[str, Any] = {}
    status_map: dict[str, Any] = {}
    overrides: dict[str, Any] = {}
    ultimo_sync: datetime | None = None
    ultimo_sync_status: str | None = None
    ultimo_sync_total: int = 0
    total_itens: int = 0


class BoardPatch(BaseModel):
    incluido: bool | None = None
    overrides: dict[str, Any] | None = None


class ResumoOut(BaseModel):
    total: int
    em_andamento: int
    atrasadas: int
    concluidas: int
    sem_responsavel: int
    sem_prazo: int
    boards: int  # quantos boards entraram na agregação


class ResponsavelRef(BaseModel):
    person_id: int | None = None
    nome: str | None = None
    kind: str | None = None


class ProjetoOut(BaseModel):
    id: str
    board_id: str
    board_nome: str
    monday_item_id: int
    nome: str | None = None
    grupo: str | None = None
    status: str | None = None
    responsaveis: list[ResponsavelRef] = []
    prazo_inicio: date | None = None
    prazo_fim: date | None = None
    concluido: bool
    atrasado: bool
    dias_atraso: int | None = None


class ProjetosPage(BaseModel):
    items: list[ProjetoOut]
    total: int
    page: int
    per_page: int


class PorResponsavelOut(BaseModel):
    person_id: int | None = None
    nome: str
    total: int
    em_andamento: int
    atrasadas: int
    concluidas: int


class TendenciaPonto(BaseModel):
    data: date
    total: int
    em_andamento: int
    atrasadas: int
    concluidas: int


class SyncResultOut(BaseModel):
    ok: bool
    boards: int = 0
    boards_ok: int = 0
    boards_erro: int = 0
    total_itens: int = 0
    detalhe: dict[str, Any] = {}

File: frontend/lib/gerencia.ts
import { api } from "@/lib/api";

// ---- tipos (espelham app/schemas/gerencia.py) ----
export type Board = {
  id: string;
  monday_board_id: number;
  nome: string;
  workspace: string | null;
  board_kind: string | null;
  incluido: boolean;
  confianca_classificacao: string | null;
  ativo: boolean;
  colunas_map: Record<string, unknown>;
  status_map: Record<string, string[]>;
  overrides: Record<string, unknown>;
  ultimo_sync: string | null;
  ultimo_sync_status: string | null;
  ultimo_sync_total: number;
  total_itens: number;
};

export type Resumo = {
  total: number;
  em_andamento: number;
  atrasadas: number;
  concluidas: number;
  sem_responsavel: number;
  sem_prazo: number;
  boards: number;
};

export type ResponsavelRef = {
  person_id: number | null;
  nome: string | null;
  kind: string | null;
};

export type Projeto = {
  id: string;
  board_id: string;
  board_nome: string;
  monday_item_id: number;
  nome: string | null;
  grupo: string | null;
  status: string | null;
  responsaveis: ResponsavelRef[];
  prazo_inicio: string | null;
  prazo_fim: string | null;
  concluido: boolean;
  atrasado: boolean;
  dias_atraso: number | null;
};

export type PorResponsavel = {
  person_id: number | null;
  nome: string;
  total: number;
  em_andamento: number;
  atrasadas: number;
  concluidas: number;
};

export type TendenciaPonto = {
  data: string;
  total: number;
  em_andamento: number;
  atrasadas: number;
  concluidas: number;
};

export type SyncResult = {
  ok: boolean;
  boards: number;
  boards_ok: number;
  boards_erro: number;
  total_itens: number;
  detalhe: Record<string, unknown>;
};

const qs = (boardId: string | null) => (boardId ? `?board_id=${boardId}` : "");

export const gerencia = {
  boards: (params?: { incluido?: boolean; ativo?: boolean }) => {
    const p = new URLSearchParams();
    if (params?.incluido !== undefined) p.set("incluido", String(params.incluido));
    if (params?.ativo !== undefined) p.set("ativo", String(params.ativo));
    const s = p.toString();
    return api.get<Board[]>(`/gerencia/boards${s ? `?${s}` : ""}`);
  },
  patchBoard: (id: string, body: { incluido?: boolean; overrides?: Record<string, unknown> }) =>
    api.patch<Board>(`/gerencia/boards/${id}`, body),
  resumo: (boardId: string | null) => api.get<Resumo>(`/gerencia/resumo${qs(boardId)}`),
  atrasadas: (boardId: string | null, limit = 100) =>
    api.get<Projeto[]>(`/gerencia/tarefas-atrasadas${qs(boardId)}${boardId ? "&" : "?"}limit=${limit}`),
  porResponsavel: (boardId: string | null) =>
    api.get<PorResponsavel[]>(`/gerencia/por-responsavel${qs(boardId)}`),
  tendencia: (boardId: string | null) =>
    api.get<TendenciaPonto[]>(`/gerencia/tendencia${qs(boardId)}`),
  sync: (boardId: string | null) => api.post<SyncResult>(`/gerencia/sync${qs(boardId)}`, {}),
};

// ---- formatação ----
export function fmtData(iso: string | null): string {
  if (!iso) return "—";
  const [y, m, d] = iso.slice(0, 10).split("-");
  if (!y || !m || !d) return "—";
  return `${d}/${m}/${y}`;
}

export function fmtDataHora(iso: string | null): string {
  if (!iso) return "nunca";
  try {
    return new Intl.DateTimeFormat("pt-BR", {
      day: "2-digit", month: "2-digit", hour: "2-digit", minute: "2-digit",
    }).format(new Date(iso));
  } catch {
    return "—";
  }
}

export function nomesResponsaveis(rs: ResponsavelRef[]): string {
  const nomes = rs.map((r) => r.nome).filter(Boolean);
  return nomes.length ? nomes.join(", ") : "—";
}
